specs/capella/beacon-chain.md

```diff
@@ -0,0 +1,299 @@
+```python
+class ExecutionPayloadHeader(Container):
+    # Execution block header fields
+    parent_hash: Hash32
+    fee_recipient: ExecutionAddress
+    state_root: Bytes32
+    receipts_root: Bytes32
+    logs_bloom: ByteVector[BYTES_PER_LOGS_BLOOM]
+    prev_randao: Bytes32
+    block_number: uint64
+    gas_limit: uint64
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    # Extra payload fields
+    block_hash: Hash32  # Hash of execution block
+    transactions_root: Root
+    withdrawals_root: Root  # [New in Capella]
+```
+
+### New containers
+
+#### `Withdrawal`
+
+```python
+class Withdrawal(Container):
+    index: WithdrawalIndex
+    address: ExecutionAddress
+    amount: Gwei
+```
+
+## Helpers
+
+### Beacon state mutators
+
+#### `withdraw`
+
+```python
+def withdraw(state: BeaconState, index: ValidatorIndex, amount: Gwei) -> None:
+    # Decrease the validator's balance
+    decrease_balance(state, index, amount)
+    # Create a corresponding withdrawal receipt
+    withdrawal = Withdrawal(
+        index=state.withdrawal_index,
+        address=state.validators[index].withdrawal_credentials[12:],
+        amount=amount,
+    )
+    state.withdrawal_index = WithdrawalIndex(state.withdrawal_index + 1)
+    state.withdrawals_queue.append(withdrawal)
+```
+
+### Predicates
+
+#### `is_fully_withdrawable_validator`
+
+```python
+def is_fully_withdrawable_validator(validator: Validator, epoch: Epoch) -> bool:
+    """
+    Check if ``validator`` is fully withdrawable.
+    """
+    is_eth1_withdrawal_prefix = validator.withdrawal_credentials[0:1] == ETH1_ADDRESS_WITHDRAWAL_PREFIX
+    return is_eth1_withdrawal_prefix and validator.withdrawable_epoch <= epoch < validator.withdrawn_epoch
+```
+
+## Beacon chain state transition function
+
+### Epoch processing
+
+```python
+def process_epoch(state: BeaconState) -> None:
+    process_justification_and_finalization(state)
+    process_inactivity_updates(state)
+    process_rewards_and_penalties(state)
+    process_registry_updates(state)
+    process_slashings(state)
+    process_eth1_data_reset(state)
+    process_effective_balance_updates(state)
+    process_slashings_reset(state)
+    process_randao_mixes_reset(state)
+    process_historical_roots_update(state)
+    process_participation_flag_updates(state)
+    process_sync_committee_updates(state)
+    process_full_withdrawals(state)  # [New in Capella]
+```
+
+#### Withdrawals
+
+*Note*: The function `process_full_withdrawals` is new.
+
+```python
+def process_full_withdrawals(state: BeaconState) -> None:
+    current_epoch = get_current_epoch(state)
+    for index, validator in enumerate(state.validators):
+        if is_fully_withdrawable_validator(validator, current_epoch):
+            # TODO, consider the zero-balance case
+            withdraw(state, ValidatorIndex(index), state.balances[index])
+            validator.withdrawn_epoch = current_epoch
```

Comment:
Yes, but `withdrawn_epoch` or any other attribute indicating that validators has been withdrawn must be set only in case of *full* withdrawal. If this would be partial withdrawal of excess balance then `withdrawn_epoch` must remain unset. This is my understanding of the design. I think we may have two functions re-using `withdraw`, each with the logic dedicated to a particular case. For instance, `withdraw_fully` may check the conditions and set `withdrawn_epoch` in addition.